src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java

```diff
@@ -110,11 +138,64 @@ private void processTripQueryParams(MonitoredTrip monitoredTrip, Request req) {
 
     @Override
     MonitoredTrip preUpdateHook(MonitoredTrip monitoredTrip, MonitoredTrip preExisting, Request req) {
-        preCreateOrUpdateChecks(monitoredTrip, req);
+        // Wait for any existing CheckMonitoredTrip jobs to complete before proceeding
+        if (monitoredTripLocks.containsKey(monitoredTrip)) {
+            int maxWaitTimeMillis = 4000;
+            int timeWaitedMillis = 0;
+            do {
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException e) {
+                    logMessageAndHalt(
+                        req,
+                        HttpStatus.INTERNAL_SERVER_ERROR_500,
+                        "The trip analyzer prevented the update of this trip. Please try again."
+                    );
+                }
+                timeWaitedMillis += 500;
 
-        // TODO: Update itinerary existence record when updating a trip.
+                // if the lock has been released, exit this wait loop
+                if (!monitoredTripLocks.containsKey(monitoredTrip)) break;
+            } while (timeWaitedMillis <= maxWaitTimeMillis);
+        }
 
-        return monitoredTrip;
+        // If a lock still exists, prevent the update
+        if (monitoredTripLocks.containsKey(monitoredTrip)) {
+            logMessageAndHalt(
+                req,
+                HttpStatus.INTERNAL_SERVER_ERROR_500,
+                "The trip analyzer prevented the update of this trip. Please try again."
+            );
+        }
+
+        // lock the trip so that the a CheckMonitoredTrip job won't concurrently analyze/update the trip.
+        monitoredTripLocks.put(monitoredTrip, true);
+
+        try {
+            preCreateOrUpdateChecks(monitoredTrip, req);
+
+            // Forbid the editing of certain values that are analyzed and set during the CheckMonitoredTrip job.
+            // For now, accomplish this by setting whatever the previous itinerary and journey state were in the preExisting
+            // trip.
+            monitoredTrip.itinerary = preExisting.itinerary;
+            monitoredTrip.journeyState = preExisting.journeyState;
+
+            checkTripCanBeMonitored(monitoredTrip, req);
+            processTripQueryParams(monitoredTrip, req);
+
+            // TODO: Update itinerary existence record when updating a trip.
+
+            // perform the database update here before releasing the lock to be sure that the record is updated in the
+            // database before a CheckMonitoredTripJob analyzes the data
+            Persistence.monitoredTrips.replace(monitoredTrip.id, monitoredTrip);
+            return runCheckMonitoredTrip(monitoredTrip);
+        } catch (Exception e) {
+            // FIXME: an error happened while checking the trip, but the trip was saved to the DB, so return the raw
+            //  trip as it was saved in the db?
+            return monitoredTrip;
```

Comment:
How do we know the exception happened while checking the trip? This try block encompasses much more than just the `runCheckMonitoredTrip` call.

Reply:
I improved the comment a bit more. This is kind of the quick and dirty way of doing things. This exception block does catch anything thrown from the runCheckMonitoredTrip job which admittedly could be a lot. However, if everything goes right, the contents of the updated trip will be returned resulting in a most-up-to-date snapshot of the trip. If we wanted to separate this, perhaps the updating could be in one webservice call and then the UI would make a second call to run the check trip update job. Or maybe we could use websockets.